woocommerce/woocommerce
# Add one-click installation for recommended extensions in multichannel Marketing page

### All Submissions:


### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/35578.

This PR depends on woocommerce.com API changes in PR https://github.com/Automattic/woocommerce.com/pull/14729.

In this PR, we add one-click installation feature to recommended extensions in Marketing page (beta multichannel view only; this PR does not affect the normal Marketing page).

In the API response, recommended extensions will have a property `direct_install` to indicate whether it supports one-click installation. `direct_install` is `true` if the extension is available in the WordPress.org plugin listings; `false` otherwise.

If `direct_install` is `true`, we will display a "Install plugin" button. Clicking it will automatically install and activate the plugin. The plugin will then be removed from the "Discover more marketing tools" card and will appear in the "Installed extensions" card.

If `direct_install` is `false`, the plugin will have a "View details" button. Clicking it will navigate to the product page on WooCommerce.com. This is same as existing behavior; we just changed the button text from "Get started" to "View details". 

The code for "automatic install and activate" is based on existing prior art in the code base, e.g.:

https://github.com/woocommerce/woocommerce/blob/0b52411ff768279dc59bb5376255c584e797970c/plugins/woocommerce-admin/client/tasks/fills/Marketing/index.tsx#L131-L156

Other technical notes: 

- I introduced a new action called `removeRecommendedPlugin`. This is called after "automatic install and activate", to remove the extension from "Discover more marketing tools".
- I added new folders `hooks` and `types` for the shared hooks and types.
- `usePlugins` has been renamed to `useInstalledPlugins` for better clarity.

📹 Demo video of the feature with my voice annotation:

https://user-images.githubusercontent.com/417342/201116343-64597380-8ac3-42ff-9b00-6441a7ee3ae3.mp4



### How to test the changes in this Pull Request:



Pre-condition: 

- Make sure you have the changes in PR https://github.com/Automattic/woocommerce.com/pull/14729.
- Enable WooCommerce Multichannel Marketing experience in WooCommerce Settings > Advanced > Features (`/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`).

Test instructions:

1. Go to Marketing page at `/wp-admin/admin.php?page=wc-admin&path=%2Fmarketing`.
2. In the "Discover more marketing tools", find a plugin with "Install plugin" button, and click on the button. 
    - The button should become busy.
    - All the other "Install plugin" and "View details" buttons should become disabled. 
    - The plugin should be installed and activated automatically.
    - There should be a successful notice message in the bottom left corner.
        - If there is an error, there should be an error notice message in the bottom left corner. You can try this out by hardcoding a non-existing plugin slug in the code.
    - The plugin should then be removed from "Discover more marketing tools" card.
    - The plugin should appear in the "Installed extensions" card.
4. In the "Discover more marketing tools", find a plugin with "View details" button, and click on the button. 
    - The browser tab should navigate to the product page on WooCommerce.com



### Other information:




### FOR PR REVIEWER ONLY:
